core/tee/tee_enc_fs.c

```diff
@@ -0,0 +1,761 @@
+
+	res = tee_fs_send_cmd(&head, NULL, 0, TEE_FS_MODE_NONE);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	return res;
+}
+
+static int read_file(int fd, void *buf, size_t len)
+{
+	int res = -1;
+	struct tee_fs_rpc head = { 0 };
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (buf == NULL) {
+		res = -1;
+		goto exit;
+	}
+
+	/* fill in parameters */
+	head.op = TEE_FS_READ;
+	head.fd = fd;
+	head.len = (uint32_t) len;
+
+	res = tee_fs_send_cmd(&head, (void *)buf, len, TEE_FS_MODE_OUT);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	return res;
+}
+
+static int write_file(int fd, const void *buf, size_t len)
+{
+	int res = -1;
+	struct tee_fs_rpc head = { 0 };
+
+	if (len == 0) {
+		res = 0;
+		goto exit;
+	}
+
+	if (buf == NULL) {
+		res = -1;
+		goto exit;
+	}
+
+	/* fill in parameters */
+	head.op = TEE_FS_WRITE;
+	head.fd = fd;
+	head.len = len;
+
+	res = tee_fs_send_cmd(&head, (void *)buf, len, TEE_FS_MODE_IN);
+	if (res)
+		goto exit;
+
+	res = head.res;
+
+exit:
+	return res;
+}
+
+static int get_file_length(int fd, size_t *length)
+{
+	size_t file_len;
+	int res;
+
+	*length = 0;
+
+	res = lseek(fd, 0, TEE_FS_SEEK_END);
+	if (res < 0)
+		return res;
+	file_len = res;
+
+	res = lseek(fd, 0, TEE_FS_SEEK_SET);
+	if (res < 0)
+		return res;
+
+	*length = file_len;
+	return 0;
+}
+
+static int update_file_size(struct tee_fs_fd *fdp,
+		size_t new_file_len)
+{
+	size_t old_file_len = fdp->len;
+
+	if (fdp->data == NULL)
+		fdp->data = malloc(new_file_len);
+	else
+		fdp->data = realloc(fdp->data, new_file_len);
+	if (fdp->data == NULL)
+		return -1;
+
+	fdp->len = new_file_len;
+
+	/* if size is increased, zero fill the gap */
+	if (new_file_len > old_file_len) {
+		size_t diff =
+			new_file_len - old_file_len;
+		memset(fdp->data + old_file_len, 0, diff);
+	}
+
+	return 0;
+}
+
+
+static int tee_fs_open(const char *file, int flags, ...)
+{
+	int nw_fd, res = -1;
+	size_t name_len;
+	size_t encrypted_data_len, file_len;
+	size_t file_header_len = tee_enc_fs_get_file_header_size();
+	void *encrypted_data;
+	TEE_Result ret;
+	struct tee_fs_fd *fd = NULL;
+	struct tee_fs_rpc head = { 0 };
+
+	DMSG("open, file=%s", file);
+
+	name_len = strlen(file) + 1;
+	if (name_len > TEE_FS_NAME_MAX)
+		goto exit;
+
+	head.op = TEE_FS_OPEN;
+	head.flags = TEE_FS_O_RDWR;
+	if (flags & TEE_FS_O_CREATE)
+		head.flags |= TEE_FS_O_CREATE;
+	head.fd = 0;
+	res = tee_fs_send_cmd(&head, (void *)file, name_len, TEE_FS_MODE_IN);
+	if (res != 0)
+		goto exit;
+
+	nw_fd = head.res;
+	if (nw_fd < 0) {
+		res = -1;
+		goto exit;
+	}
+
+	fd = (struct tee_fs_fd *)malloc(sizeof(struct tee_fs_fd));
+	if (fd == NULL) {
+		res = -1;
+		goto exit;
+	}
+
+	res = get_file_length(nw_fd, &encrypted_data_len);
+	if (res < 0) {
+		res = -1;
+		goto exit;
+	}
+
+	/* new created file? */
+	if (encrypted_data_len == 0) {
+		fd->nw_fd = nw_fd;
+		fd->flags = flags;
+		fd->pos = fd->len = 0;
+		fd->data = NULL;
+		return handle_get(&fs_handle_db, fd);
```

Comment:
The allocated 'fd' is freed at 'close()'. It's the caller's('tee_svc_storage.c') responsibility to handle this, if you open a file and get a file handle, you are supposed to close it somewhere.